### PR TITLE
Fixing how subqueries use keys when building

### PR DESCRIPTION
**Pull request by patrickconroy**
Refs #6781.

### CODE REVIEW

**Comment by ADmad** [comment]
Can you provide a test case please.

**Comment by patrickconroy** [comment]
TIL postgres sorts data weirdly.. fixing that now.

**Comment by patrickconroy** [comment]
The fail looks unrelated now. I don't know how I'd fix this..

``
Fatal error: Uncaught exception 'Cake\Core\Exception\Exception' with message 'Unable to insert fixtures for "Cake\Test\TestCase\ORM\QueryRegressionTest" test case. SQLSTATE[23000]: [Microsoft][SQL Server Native Client 11.0][SQL Server]Cannot insert explicit value for identity column in table 'featured_tags' when IDENTITY_INSERT is set to OFF.' in C:\projects\cakephp\src\TestSuite\Fixture\FixtureManager.php:254
Stack trace:
``

**Comment by lorenzo** [comment]
@patrickconroy it means that you are trying to insert a record having a known primary key, instead of letting the database create the next auto incremental id for you.

File: tests/Fixture/TagsTranslationsFixture.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+class TagsTranslationsFixture extends TestFixture
+{
+
+    /**
+     * fields property
+     *
+     * @var array
+     */
+    public $fields = [
+        'id' => ['type' => 'integer', 'null' => false],
```
Comment (lorenzo) [inline tests/Fixture/TagsTranslationsFixture.php]: I *think* you will need to add `'autoIncrement' => false` here, although I'm not sure

**Comment by patrickconroy** [comment]
Hm, I'm not doing that. The fixture has a `records` array. There's no auto incremental id since the PK is id+locale. How/why would it attempt to increment a string?

File: tests/Fixture/TagsTranslationsFixture.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        '_constraints' => ['primary' => ['type' => 'primary', 'columns' => ['id', 'locale']]]
+    ];
+
+    /**
+     * records property
+     *
+     * @var array
+     */
+    public $records = [
+        ['id' => '1', 'locale' => 'en_us', 'name' => 'tag 1 translated into en_us'],
```
Comment (ADmad) [inline tests/Fixture/TagsTranslationsFixture.php]: Remove id from records and the postgres issue should be solved.

Comment (patrickconroy) [inline tests/Fixture/TagsTranslationsFixture.php]: How would finding these make sense, then? The index in that array of those records !== the id..

Comment (ADmad) [inline tests/Fixture/TagsTranslationsFixture.php]: The value for `id` field will be auto inserted by db. I don't think the `'autoIncrement' => false` @lorenzo suggested is required.

Comment (lorenzo) [inline tests/Fixture/TagsTranslationsFixture.php]: He has duplicated `'id'` values @ADmad

Comment (patrickconroy) [inline tests/Fixture/TagsTranslationsFixture.php]: The `id` value can't be auto-inserted, it doesn't make sense that way. There could be multiple `id=1`s or there could be 300 `id=17`s, it's a composite PK based on the id + locale. Maybe I'm missing something really simple about what you're saying though..

Comment (ADmad) [inline tests/Fixture/TagsTranslationsFixture.php]: Error says: "**Cannot insert explicit value for identity column** in table 'featured_tags' when IDENTITY_INSERT is set to OFF".  Specifying the 'id' is trying to insert explicit value isn't it?

That said I have no experience with postgres, but have encountered similar error in the past (iirc) and removing ids from records fixed it.

Comment (ADmad) [inline tests/Fixture/TagsTranslationsFixture.php]: Apologies, please ignore me. Like I said my suggestion was based off something i remembered to have encountered and not actual postgres experience :)

Comment (ADmad) [inline tests/Fixture/TagsTranslationsFixture.php]: And I am an idiot too, the error is for ms sql not postgres.

Comment (patrickconroy) [inline tests/Fixture/TagsTranslationsFixture.php]: Well, it looks like an ms sql error and not postgres, right? Are there other fixtures that use a non-id, non-autoincrement PK?

Comment (patrickconroy) [inline tests/Fixture/TagsTranslationsFixture.php]: Hah no worries. The postgres stuff I got around another way; apparently the default sort on the records with a composite PK are opposite there.

**Comment by patrickconroy** [comment]
So at this point I'm not sure how to fix that. Does anyone have any ideas? Is this an issue with how fixtures in ms sql work?

File: tests/Fixture/FeaturedTagsFixture.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * @link          http://book.cakephp.org/2.0/en/development/testing.html CakePHP(tm) Tests
+ * @since         1.2.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Test\Fixture;
+
+use Cake\TestSuite\Fixture\TestFixture;
+
+/**
+ * Class TagFixture
```
Comment (markstory) [inline tests/Fixture/FeaturedTagsFixture.php]: Lies :smile:

Comment (patrickconroy) [inline tests/Fixture/FeaturedTagsFixture.php]: My copy/paste game wasn't great today I guess.

File: tests/Fixture/FeaturedTagsFixture.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * CakePHP(tm) Tests <http://book.cakephp.org/2.0/en/development/testing.html>
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://book.cakephp.org/2.0/en/development/testing.html CakePHP(tm) Tests
+ * @since         1.2.0
```
Comment (ADmad) [inline tests/Fixture/FeaturedTagsFixture.php]: Lies

**Comment by lorenzo** [comment]
@markstory Do you happen to know how to make the fixture not create an auto increment id?

File: tests/Fixture/FeaturedTagsFixture.php (hunk 4 above)
Comment (patrickconroy) [inline tests/Fixture/FeaturedTagsFixture.php]: Should I make that 3.0.7 ? I obviously don't know if this will make it to that, though.

**Comment by patrickconroy** [comment]
I changed the fixture to not have set primary keys in the records. It seems weird that I had to.. but it looks like that fixed it. The important thing to know is that the test I added would fail if I didn't update `SelectableAssociationTrait::_buildSubquery`.

**Comment by markstory** [comment]
@lorenzo you can set `autoIncrement = false`. If you need/want an autoincrement you will need to omit the autoIncrement columns from the fixture to make sqlserver happy.

**Comment by markstory** [comment]
@patrickconroy looks good. Could you squash out the 'try this' commits?

**Comment by lorenzo** [comment]
Squashed and merged in master. Thanks @patrickconroy !

**Comment by patrickconroy** [comment]
Thanks for doing that! I'm afk until tomorrow.

**Pull request closed without merging**